### ISSUE TITLE
cumsum drops index coordinates

### ISSUE THREAD

**Issue by dcherian** [pydata/xarray#6528]
### What happened?

cumsum drops index coordinates. Seen in #6525, #3417

### What did you expect to happen?

Preserve index coordinates

### Minimal Complete Verifiable Example

```Python
import xarray as xr

ds = xr.Dataset(
        {"foo": (("x",), [7, 3, 1, 1, 1, 1, 1])},
        coords={"x": [0, 1, 2, 3, 4, 5, 6]},
    )
ds.cumsum("x")
```

```
<xarray.Dataset>
Dimensions:  (x: 7)
Dimensions without coordinates: x
Data variables:
    foo      (x) int64 7 10 11 12 13 14 15
```

### Environment

<details>

xarray main

</details>

**Comment by Illviljan** [comment]
https://github.com/pydata/xarray/blob/7173bd3f6b76e92f77f4156236fd2b3fe546c9c2/xarray/core/dataset.py#L5239-L5240
If you comment out this if-check so it always add the variable it seems to work. 

Not sure how to handle this:
* Add a new keyword argument to `reduce`? `keepcoords` or can `keepdims` be expanded to handle this as well?
* Since `cumsum` is not really a reduction, should we create a new method instead of `reduce`? I guess it would very similar to reduce though.

**Comment by dcherian** [comment]
> Since cumsum is not really a reduction, should we create a new method instead of reduce? I guess it would look very similar to reduce though.

:+1: I think we should add cumsum to `generate_reductions.py`. which maybe should become `generate_aggregations.py`

**Comment by max-sixty** [comment]
There was some discussion a while ago about how to handle `reduce` on things that aren't reductions. There's some overlap with https://github.com/pydata/xarray/issues/1618, and that has some links to other issues.

**Comment by max-sixty** [comment]
Was this closed by https://github.com/pydata/xarray/pull/7152?

The general reduce issue overlaps with https://github.com/pydata/xarray/issues/6528.

**Comment by max-sixty** [comment]
I had a look at moving this to `.apply_ufunc` — if I'm thinking about it correctly — it's actually not trivial, since it's not just a gufunc like the `.rolling_exp` ones from #8114.

Instead, it's a function that we've written which can execute over multiple dimensions https://github.com/pydata/xarray/blob/c9d89e2b51b6dfee34e371a9232024ec45cd2d9c/xarray/core/duck_array_ops.py#L621-L631. (not super difficult! But IIUC, less of a lift-and-shift...)

**Comment by max-sixty** [comment]
Another alternative is to focus on `Cumulative` — `ds.cumulative('lat').sum()` works great!

**Comment by dcherian** [comment]
Can we just redirect `cumsum` to that then?

**Comment by max-sixty** [comment]
> Can we just redirect `cumsum` to that then?

I looked at this but it has slightly more features such as `skipna`...

**Comment by jhollowed** [comment]
> Can we just redirect `cumsum` to that then?

fwiw, I just tested this with a 33 MB `Dataset`. I observed that

`data.cumsum('time')`

is anywhere from 18 to 30x faster than

`data.cumulative('time').sum()`

**Comment by max-sixty** [comment]
@jhollowed, is that the first invocation? `.cumulative` is JIT compiled with numba, so the first invocation of the session will be much slower. But then it should be between a bit faster and _much_ faster if it can parallelize.

If you have a repro then happy to investigate, would be a big deal if that's real.

**Comment by jhollowed** [comment]
@max-sixty No, it's not just the first invocation, from what I can tell. The repo in question is a mess, but here's a minimal working example which mimics my use case:

```
import time
import numpy as np
import xarray as xr
import matplotlib.pyplot as plt

data = xr.DataArray(np.random.rand(100,10,1000), dims=('x','y','z'))

cumsum_time = np.zeros(len(data.z))
cumulative_time = np.zeros(len(data.z))

for i in range(100):
    start = time.time()
    data.cumsum('z')
    cumsum_time[i] = time.time() - start

    start = time.time()
    datac.sum()
    cumulative_time[i] = time.time() - start

plt.hist(cumulative_time/cumsum_time, bins=20)
plt.show()
```

In this example the difference is even more extreme, with a ~50x slowdown by using `cumulative()`.

<img width="598" alt="Screenshot 2024-10-07 at 2 25 05 PM" src="https://github.com/user-attachments/assets/3d7c6db0-947b-409f-a9ae-d20ac7c1554e">

I tried to also save the return of `data.cumulative('z')` as a new object prior to the loop, and call only `sum()` within the loop. This made no qualitative difference to the timing result. I also tried this in a Jupyter notebook, in which the difference was even worse but still qualitatively similar. I am running xarray version 2024.6.0.

This is perhaps veering off-topic for this issue, but long story short, I just ended up using `cumsum()` and `assign_coords()` to recover the index coordinate dropped by `cumsum()`, rather than using the much slower `cumulative().sum()`.

**Comment by max-sixty** [comment]
Can you share your env? I get something very very different!

<img width="726" alt="image" src="https://github.com/user-attachments/assets/185f8f0d-cf91-4224-9187-c01055a9e6fd">

The one outlier will be the initial JIT run

**Comment by max-sixty** [comment]
FWIW you may also want to use something like `%%timeit` — the repro above is using a missized array to store the values etc (no stress, I'd have done the same, but that's why `%%timeit` is great)

**Comment by jhollowed** [comment]
@max-sixty Interesting. 

`timeit.timeit(lambda: data.cumsum('z'), number=100)`

and 

`timeit.timeit(lambda: data.cumulative().sum('z'), number=100)`

yield `1.056` and `50.75`, respectively.

I'm working on a login node of a cluster with many users, so I'm not totally in control of the context I'm running in. But my conda env looks like [this](https://pastebin.com/2AKKGxfH).

**Comment by max-sixty** [comment]
Ah, OK, neither `numbagg` nor `bottleneck` is installed, and so it defaults to the slow routine. I guess that's more of a trap than I realized, I should have thought of that initially. We do at least put that in the `.cumulative` docstring.

FWIW we're considering including `numbagg` in the standard xarray release to reduce this confusion, and having a "core" release without any optional dependencies. 

In the meantime, you might want to add `numbagg` to the env. (disclaimer: also a maintainer of numbagg...)

**Comment by max-sixty** [comment]
Given we now direct folks to `.cumulative`, will plan to close this.

I thought about whether we should add a warning for `.cumulative` without `numbagg` nor `bottleneck`; I think that's a bit too invasive, and will be screened off if we do the `xarray-core` package. But open to feedback ofc.

**Comment by dcherian** [comment]
The core `cumsum` is still buggy

**Comment by max-sixty** [comment]
> The core `cumsum` is still buggy

Agree; I closed since it seems unlikely we'll fix. But great if there's a path.

We could deprecate the `skipna`, and then defer to `.cumulative`. But requires numbagg...
